commit 093d1bbfb744bf685c722652afc1be2118b211f2 (refs/changes/09/1509/8)
Author: Nick Zivkovic <nick.zivkovic@joyent.com>
Date:   2018-04-02T20:27:22+00:00 (1 year, 6 months ago)
    
    NAPI-326 'free' and 'unassign' IP need their usages untangled
    Reviewed by: Cody Peter Mello <cody.mello@joyent.com>
    Approved by: Cody Peter Mello <cody.mello@joyent.com>

diff --git a/lib/endpoints/networks/ips.js b/lib/endpoints/networks/ips.js
index 38fde2f..094b81a 100644
--- a/lib/endpoints/networks/ips.js
+++ b/lib/endpoints/networks/ips.js
@@ -18,6 +18,7 @@ var mod_common = require('./common');
 var mod_ip = require('../../models/ip');
 var restify = require('restify');
 var util_ip = require('../../util/ip');
+var mod_jsprim = require('jsprim');
 
 
 
@@ -111,7 +112,9 @@ function putIP(req, res, next) {
     // the mod_ip.* functions require a network model object:
     req.params.network = req._network;
 
-    if (req.params.hasOwnProperty('free') && req.params.free) {
+    if (mod_jsprim.hasKey(req.params, 'free') && req.params.free &&
+        !mod_jsprim.hasKey(req.params, 'unassign')) {
+
         return mod_ip.del(req.app, req.log, req.params, function (err) {
             if (err && err.statusCode !== 404) {
                 return next(err);
diff --git a/lib/models/ip/index.js b/lib/models/ip/index.js
index 35b9a67..abe2a54 100644
--- a/lib/models/ip/index.js
+++ b/lib/models/ip/index.js
@@ -26,7 +26,6 @@ var util_ip = require('../../util/ip');
 var validate = require('../../util/validate');
 
 
-
 // --- Globals
 
 
@@ -143,7 +142,19 @@ function validateNetworkOwner(_opts, _, validated, callback) {
             constants.OWNER_MATCH_MSG));
     }
 
-    return callback();
+    callback();
+}
+
+/*
+ * The `free` and `unassign` property can't be set to `true` at the same time.
+ */
+function validateFreeUnassign(_opts, _, validated, callback) {
+    if (validated.free && validated.unassign) {
+        return callback(errors.invalidParam('unassign',
+            constants.FREE_UNASSIGN_MSG));
+    }
+
+    callback();
 }
 
 
@@ -316,14 +327,16 @@ function updateIP(app, log, params, callback) {
             belongs_to_uuid: validate.UUID,
             check_owner: validate.bool,
             owner_uuid: validate.UUID,
-            reserved: validate.bool
+            reserved: validate.bool,
+            unassign: validate.bool,
+            free: validate.bool
         },
 
         required: {
             network: validateNetworkObj
         },
 
-        after: validateNetworkOwner
+        after: [validateNetworkOwner, validateFreeUnassign]
     };
 
     // both belongs_to_type and belongs_to_uuid must be set in UFDS at the
@@ -366,7 +379,7 @@ function updateIP(app, log, params, callback) {
 
         // If unassigning, remove the 'belongs_to' information, but keep
         // owner and reserved
-        if (params.unassign) {
+        if (validatedParams.unassign) {
             updateOpts.val = {
                 belongs_to_type: true,
                 belongs_to_uuid: true
diff --git a/lib/util/constants.js b/lib/util/constants.js
index 74aef90..0f433f8 100644
--- a/lib/util/constants.js
+++ b/lib/util/constants.js
@@ -164,6 +164,7 @@ module.exports = {
     MTU_NICTAG_UPDATE_MSG: 'nic_tag mtu update must support existing networks',
     NIC_PROVISION_RETRIES: 100,
     OWNER_MATCH_MSG: 'network owner_uuids do not match',
+    FREE_UNASSIGN_MSG: 'free and unassign cannot both be set to "true"',
     POOL_IP_MSG: 'IP cannot be specified with a network pool',
     POOL_MIN_NETS_MSG:
         'network pool must contain at least one network',
diff --git a/test/unit/ips.test.js b/test/unit/ips.test.js
index 26077dd..095ba17 100644
--- a/test/unit/ips.test.js
+++ b/test/unit/ips.test.js
@@ -26,6 +26,7 @@ var mod_uuid = require('node-uuid');
 var test = require('tape');
 var util = require('util');
 var vasync = require('vasync');
+var constants = require('../../lib/util/constants');
 
 
 
@@ -716,6 +717,26 @@ test('Update IP - free (IP not in moray)', function (t) {
     });
 });
 
+test('Update IPv4 - unassign and free (IP in moray)', function (t) {
+    var params = {
+        belongs_to_type: 'server',
+        belongs_to_uuid: mod_uuid.v4(),
+        owner_uuid: mod_uuid.v4()
+    };
+
+    NAPI.updateIP(NETV4.uuid, '10.0.2.34', params, function (err) {
+        t.ifError(err);
+        var uf_params = { unassign: 'true', free: 'true' };
+        NAPI.updateIP(NETV4.uuid, '10.0.2.34', uf_params,
+            function (err2, _, req, res) {
+                t.ok(err2, 'Expecting error');
+                t.deepEqual(err2.body.errors[0], {field: 'unassign',
+                    code: 'InvalidParameter',
+                    message: constants.FREE_UNASSIGN_MSG});
+                t.end();
+            });
+    });
+});
 
 test('Update IPv4 - unassign (IP in moray)', function (t) {
     var params = {
